src/dsa/rpo_falcon512/signature.rs

```diff
@@ -39,251 +41,332 @@ use super::{
+        Self(pk_poly)
     }
+}
+
+impl TryFrom<&[i16; N]> for SignaturePoly {
+    type Error = ();
 
-    // squeeze the coefficients of the polynomial
-    let mut i = 0;
-    let mut res = [0_u16; N];
-    for _ in 0..64 {
-        Rpo256::apply_permutation(&mut state);
-        for a in &state[Rpo256::RATE_RANGE] {
-            res[i] = (a.as_int() % MODULUS as u64) as u16;
-            i += 1;
+    fn try_from(coefficients: &[i16; N]) -> Result<Self, Self::Error> {
+        if are_coefficients_valid(coefficients) {
+            Ok(Self(coefficients.to_vec().into()))
+        } else {
+            Err(())
         }
     }
-
-    // using the raw constructor is OK here because we reduce all coefficients by the modulus above
-    unsafe { Polynomial::new(res) }
 }
 
-/// Converts byte representation of the nonce into field element representation.
-fn decode_nonce(nonce: &NonceBytes) -> NonceElements {
-    let mut buffer = [0_u8; 8];
-    let mut result = [ZERO; 8];
-    for (i, bytes) in nonce.chunks(5).enumerate() {
-        buffer[..5].copy_from_slice(bytes);
-        // we can safely (without overflow) create a new Felt from u64 value here since this value
-        // contains at most 5 bytes
-        result[i] = Felt::new(u64::from_le_bytes(buffer));
+impl Serializable for &SignaturePoly {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let sig_coeff: Vec<i16> = self.0.coefficients.iter().map(|a| a.balanced_value()).collect();
+        let sk_bytes = compress_signature(&sig_coeff);
+        target.write_bytes(&sk_bytes);
     }
+}
 
-    result
+impl Deserializable for SignaturePoly {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let bytes = source.read_array::<SIG_LEN>()?;
+        decompress_signature(&bytes)
+            .map_err(|err| DeserializationError::InvalidValue(err.to_string()))
+            .map(SignaturePoly::from)
+    }
 }
 
-// TESTS
+// HELPER FUNCTIONS
 // ================================================================================================
 
-#[cfg(all(test, feature = "std"))]
-mod tests {
-    use core::ffi::c_void;
-    use rand_utils::rand_vector;
-
-    use super::{
-        super::{ffi::*, KeyPair},
-        *,
-    };
-
-    // Wrappers for unsafe functions
-    impl Rpo128Context {
-        /// Initializes the RPO state.
-        pub fn init() -> Self {
-            let mut ctx = Rpo128Context { content: [0u64; 13] };
-            unsafe {
-                rpo128_init(&mut ctx as *mut Rpo128Context);
-            }
-            ctx
-        }
+/// Takes the hash-to-point polynomial `c` of a message and the signature polynomials over
+/// the message `(s1, s2)` and returns `true` is the signature is a valid signature for
+/// the given parameters, otherwise it returns `false`.
+fn verify_helper(
+    c: Polynomial<FalconFelt>,
+    s1: SignaturePoly,
+    s2: SignaturePoly,
+    pubkey_com: Word,
+) -> bool {
```

Comment:
I think this method should take `PubKeyPoly` instead of `s1` and `pubkey_com` as parameters. The main idea is that we'll be able to reuse it for canonical signature verification as well.

When called from `Signature::verify()` method, we can first compute the `h` polynomial in `Signature::verify()`, check it against the public key commitment, and then pass it here.

Reply:
Yes, I have updated it to the above but this would entail a few redundant operations, maybe worth it for testing the canonical implementation though.

Reply:
Yeah - that's what I was thinking: we give up a bit on performance side, but would be able to test more against the canonical implementation.